### PR TITLE
Bandwidth limited diffusion

### PR DESCRIPTION
**Pull request by cskiraly**
This PR changes simulation granularity from lines (rows/columns) to segments, and introduces bandwidth limitations.

### CODE REVIEW

File: DAS/validator.py [hunk 1]
```diff
@@ -66,6 +134,22 @@ def __init__(self, ID, amIproposer, logger, shape, rows, columns):
...
+        self.repairOnTheFly = True
+        self.sendLineUntil = (self.shape.blockSize + 1) // 2 # stop sending on a p2p link if at least this amount of samples passed
+        self.perNeighborQueue = True # queue incoming messages to outgoing connections on arrival (as typical GossipSub impl)
+        self.shuffleQueues = True # shuffle the order of picking from active queues of a sender node
+        self.perNodeQueue = False # keep a global queue of incoming messages for later sequential dispatch
+        self.shuffleLines = True # shuffle the order of rows/columns in each iteration while trying to send
+        self.shuffleNeighbors = True # shuffle the order of neighbors when sending the same segment to each neighbor
+        self.dumbRandomScheduler = False # dumb random scheduler
+        self.segmentShuffleScheduler = True # send each segment that's worth sending once in shuffled order, then repeat
+        self.segmentShuffleSchedulerPersist = True # Persist scheduler state between timesteps
```
Comment (leobago) [inline DAS/validator.py:151]: Is there any realistic case in which we want this not to persist? 
Just curiosity, is good to have it as a parameter as well.

File: DAS/validator.py [hunk 2]
```diff
@@ -3,24 +3,92 @@
...
+
+class NextToSend:
+    def __init__(self, neigh, toSend, id, dim):
+        self.neigh = neigh
+        self.toSend = toSend
+        self.id = id
+        self.dim = dim
+
+class SegmentToSend:
+    def __init__(self, dim, id, i):
```
Comment (leobago) [inline DAS/validator.py:72]: What is i here?

File: DAS/validator.py [hunk 3]
```diff
@@ -174,48 +256,183 @@ def updateStats(self):
...
+                if dim == 0:
+                    self.sendSegmentToNeigh(lineID, neigh.sendQueue.popleft(), neigh)
+                else:
+                    self.sendSegmentToNeigh(neigh.sendQueue.popleft(), lineID, neigh)
+                progress = True
+                if self.statsTxInSlot >= self.bwUplink:
+                    return
+
+    def runSegmentShuffleScheduler(self):
+            # This scheduler check which owned segments needs sending (at least
```
Comment (leobago) [inline DAS/validator.py:321]: Would be nice to catch this in the doc

**Review by leobago — COMMENTED** [review]
I think we can leave the configuration changes for the next PR and leave the current hard-coded parameters as they are now. But please check the other comments, most of them are minor but necessary to improve the readability of the code.

File: DAS/block.py [hunk 4]
```diff
@@ -20,6 +20,12 @@ def merge(self, merged):
         """It merges (OR) the existing block with the received one."""
         self.data |= merged.data
 
+    def getSegment(self, rowID, columnID):
+        return self.data[rowID*self.blockSize + columnID]
+
+    def setSegment(self, rowID, columnID, v = 1):
```
[the remark this answers is not in the document]
Comment (cskiraly) [reply DAS/block.py:26]: right, fixing

File: DAS/block.py [hunk 5]
```diff
@@ -20,6 +20,12 @@ def merge(self, merged):
         """It merges (OR) the existing block with the received one."""
         self.data |= merged.data
 
+    def getSegment(self, rowID, columnID):
```
[the remark this answers is not in the document]
Comment (cskiraly) [reply DAS/block.py:23]: Checking the code, I would opt for calling the elementary unit in the block a segment. We only use sample in a few places, and "sample" is already overloaded as random.sample, sampling, etc.

File: DAS/validator.py [hunk 6]
```diff
@@ -3,24 +3,92 @@
...
 import logging
+import sys
 from DAS.block import *
 from bitarray import bitarray
 from bitarray.util import zeros
-
+from collections import deque
+from itertools import chain
+
+def shuffled(lis, shuffle=True):
```
[the remark this answers is not in the document]
Comment (cskiraly) [reply DAS/validator.py:13]: right, moving it out

File: DAS/validator.py (hunk 1 above)
Comment (cskiraly) [reply DAS/validator.py:151]: if you persist between steps, you don't work with fresh information. On the practical side, persist seems to be a better model.

File: DAS/validator.py (hunk 2 above)
Comment (cskiraly) [reply DAS/validator.py:72]: i is the position in the line. Renaming to dim, lineId, id

**Pull request merged**